src/wasm/wat-parser.cpp

```diff
@@ -0,0 +1,394 @@
+/*
+ * Copyright 2022 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cctype>
+#include <optional>
+#include <variant>
+
+#include "wat-parser.h"
+
+using namespace std::string_view_literals;
+
+namespace wasm::WATParser {
+
+namespace {
+
+// ================
+// Lexical Analysis
+// ================
+
+struct LexCtx {
+protected:
+  std::string_view original;
```

Comment:
It's a slice into the full input buffer that contains the unlexed input at the time the `LexCtx` is constructed. Logically it should be a pair of `string_view`s, "lexed" and "next", but if we stored them like that, we wouldn't be able to grow "lexed" because there's no way to grow a `string_view`. So instead we keep track of the boundary point.

Reply:
IIUC, then if the full input is `ABCD` then `original` in the first `LexCtx` is `ABCD`, then later we have a `LexCtx` with `BCD`, then `CD`, and then `D`?

Reply:
Or is the idea there will always ever be a single `LexCtx`, and only `index` is incremented?

Reply:
I'm doing some confusing magic here where all the fragment lexers take `LexCtx` as their argument, but actually they're only ever called with `std::string_view` and I depend on an implicit conversion to construct the `LexCtx`. I'm about to push a change that makes this much clearer by making each fragment lexer take a `string_view` argument and explicitly construct a `LexCtx` from it. Hopefully that will make things clearer.